Escape textual ion symbols less often

joshterrell805 opened the pull request:
I didn't find a definitive list of keywords. I'm not sure if there are any others besides `null, nan, false, true`. I'd expect `inf` to be a keyword, but it worked out okay without quoting.

Found this: https://github.com/amzn/ion-python/blob/master/amazon/ion/reader_text.py#L1279-L1289

joshterrell805 commented on tests/writer_util.py:
```diff
@@ -60,9 +60,14 @@ def _scalar_p(ion_type, value, expected, force_stream_end):
+        if literal.decode('utf-8')[0] == "'":
+            yield (value, literal.replace(b"'", b'"'))
+        else:
+            # there's one symbol in our symbol table which isn't quoted
```
That makes sense, silly me :)

joshterrell805 commented on tests/test_writer_text.py:
```diff
@@ -237,3 +239,53 @@ def test_valid_indent_strs(p):
+        _P_Q('\'', True, backslash_required=True),
+        _P_Q('"', True),
+        _P_Q('\\', True, backslash_required=True),
+        # begin reserved words
```
Alright, I'll add them as well

joshterrell805 commented on tests/test_simpleion.py:
```diff
@@ -507,16 +508,21 @@ class PrettyPrintParams(record('ion_text', 'indent', ('exact_text', None), ('nee
+        PrettyPrintParams(ion_text='[apple, "banana", {roof: false}]', indent='\t',
+            exact_text="$ion_1_0\n[\n\tapple,\n\t\"banana\",\n\t{\n\t\troof: false\n\t}\n]"),
         PrettyPrintParams(ion_text='[apple, {roof: false, walls:4, door: wood::large::true}]', indent='\t',
-            needles=["$ion_1_0\n[\n\t'apple',\n\t{", "\n\t}\n]"],
```
Oh weird, looks like I accidentally removed this. Yeah if it's just being used for one test case, I'll remove the needles.

joshterrell805 commented on amazon/ion/writer_text.py:
```diff
@@ -241,19 +241,23 @@ def _bytes_text(code_point_iter, quote, prefix=b'', suffix=b''):
 _SINGLE_QUOTE = b"'"
 _DOUBLE_QUOTE = b'"'
+_UNQUOTED_SYMBOL_REGEX = re.compile(r'\A[a-zA-Z$_][a-zA-Z0-9$_]*\Z')
+_SYMBOLS_REQUIRING_QUOTES = ['nan', 'null', 'false', 'true']
```
Hmm yes faster, more to maintain, but I think clearer perhaps. I'll take another pass at this. Thanks for the feedback!

raganhan commented:
@joshterrell805 can you confirm if you made the changes that were asked by @tgregg? It seems like you did since the some comments are marked as outdated and all were marked as resolved, but there is only one commit in the PR history. 

Just want to double check before merging the change

joshterrell805 commented:
Hi @raganhan, I addressed all comments, but I pushed a forced revision which is why there's only one commit. For the following comment:

> The other reserved words are +inf, -inf, and all of the typed nulls, e.g. null.int. However, all of these are obviously flagged by _UNQUOTED_SYMBOL_REGEX, so the code is still correct as-is. Do you think it's faster to look them up from _SYMBOLS_REQUIRING_QUOTES?

Instead of adding those symbols to the `_SYMBOLS_REQUIRING_QUOTES`, list, I renamed that list and added a comment making the semantics clearer.

raganhan commented:
@joshterrell805 sounds good, merging 

Thanks for the contribution :)
